### PR TITLE
[Testcase Refactoring] Make test_serialize.py device-agnostic via instantiate + device

### PR DESCRIPTION
**Pull request by SHIHAIFENG15**
## Summary
Make `test/export/test_serialize.py` device-agnostic via `instantiate_device_type_tests` + injected `device`, so accelerators (including privateuse1/NPU) are covered without using `HAS_GPU` / `GPU_TYPE` as the entry path.

## Changes
- Convert relevant serialize tests to `instantiate_device_type_tests` with injected `device`.
- Add `hw_classification` where needed.
- Gate Triton tests with a local `_gate_triton(device)`:
  - skip only if the Triton package / imports are missing (`has_triton_package()` / `triton` / `wrap_triton`);
  - then call `get_interface_for_device(...).raise_if_triton_unavailable(device)` for a **per-device** check.

## Test Plan

```bash
python test/export/test_serialize.py -v -k TestSerializeTriton
python test/export/test_serialize.py -v -k TestSerializeAccelerator
```

### CODE REVIEW

**Comment by SHIHAIFENG15** [comment]
Update: This PR is fully labeled (hw_classification on serialize split classes) and TestSerializeTriton now uses @requires_capabilities(Capability.lib.triton) instead of _gate_triton() calling has_triton() — labeling and capability-gating work here are done.

Blocked on #190846: CI won’t pass until the capability infrastructure lands. No need to re-run CI on this update for now.

File: test/export/test_serialize.py [hunk 1]
```diff
@@ -2894,5 +2610,312 @@ def forward(self, x, y):
...
+
+        triton_node = None
+        for node in serialized.exported_program.graph_module.graph.nodes:
+            if node.target == "torch.ops.higher_order.triton_kernel_wrapper_functional":
+                triton_node = node
+                break
+        self.assertIsNotNone(triton_node)
+
+
+instantiate_device_type_tests(TestSerializeAccelerator, globals(), except_for="cpu")
```
Comment (lizhihai137) [inline test/export/test_serialize.py]: The MPS and XPU device types are not covered.

**Review by lizhihai137 — CHANGES_REQUESTED** [review]
TestSerializeAccelerator Test Case：The MPS and XPU device types are not covered.

File: test/export/test_serialize.py (hunk 1 above)
Comment (SHIHAIFENG15) [reply test/export/test_serialize.py]: Thanks — updated the three non-Triton accelerator suites to pass allow_mps=True and allow_xpu=True.

**Review by lizhihai137 — APPROVED** [review]
Comments have been processed.

**Review by fffrog — CHANGES_REQUESTED** [review]
Please waiting this [PR](https://github.com/pytorch/pytorch/pull/190846) merged first, thank you for your patience.

**Review by lizhihai137 — APPROVED** [review]
thanks for your contribution

**Comment by SHIHAIFENG15** [comment]
hat PR landed without Capability.lib.triton (only dtype.* / attention.*), so
@requires_capabilities(Capability.lib.triton) is not valid on current main —
the capability is undeclared and the test would AssertionError instead of skip.
I dropped that decorator. TestSerializeTriton now:
- skips only if the triton package is missing (has_triton_package() / import)
- looks up DeviceInterface for the injected device via
  get_interface_for_device(device)
- does not use process-level has_triton(), which is "any known device in this
  process", not "this test's device"

**Review by lizhihai137 — APPROVED** [review]

**Review by lizhihai137 — APPROVED** [review]

File: test/export/test_serialize.py [hunk 2]
```diff
@@ -2928,5 +2609,317 @@ def forward(self, x, y):
...
+            self.assertEqual(
+                param.device.type,
+                torch.device(device).type,
+                lambda msg: f"{msg}\n{name} not on {device}",
+            )
+        self.assertEqual(m(*inp), loaded_ep.module()(*inp))
+
+
+@unittest.skipUnless(HAS_TRITON, "requires triton")
+class TestSerializeTriton(TestCase):
```
Comment (fffrog) [inline test/export/test_serialize.py]: Please move those testcases in `TestSerializeTriton` into `TestSerializeAccelerator` with the decorator `@unittest.skipUnless(HAS_TRITON, "requires triton")`

File: test/export/test_serialize.py [hunk 3]
```diff
@@ -2928,5 +2609,317 @@ def forward(self, x, y):
...
+    except_for="cpu",
+    allow_mps=True,
+    allow_xpu=True,
+)
+instantiate_device_type_tests(
+    TestSaveLoadAccelerator, globals(), except_for="cpu", allow_mps=True, allow_xpu=True
+)
+instantiate_device_type_tests(
+    TestSerializeTriton, globals(), except_for="cpu", allow_xpu=True
+)
```
Comment (fffrog) [inline test/export/test_serialize.py]: Remove this one

**Review by fffrog — APPROVED** [review]
SGTM

File: test/export/test_serialize.py (hunk 2 above)
Comment (SHIHAIFENG15) [reply test/export/test_serialize.py]: Thanks, addressed. Folded `test_triton_hop` and `test_triton_constexpr_matching` into `TestSerializeAccelerator` with `@unittest.skipUnless(HAS_TRITON, "requires triton")`, and removed `TestSerializeTriton` (including its `instantiate_device_type_tests`). Those two methods are also skipped on MPS because `TestSerializeAccelerator` is instantiated with `allow_mps=True` (for weight sharing), and Triton is not available on MPS.

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill

**Comment by SHIHAIFENG15** [comment]
> **Claude finished @can-gaa-hou's task in 13m 29s** —— [View job](https://github.com/pytorch/pytorch/actions/runs/33827580198)
> 
> The refactor's direction is right, but the `HAS_TRITON` gate is process-level rather than per-device, so the two Triton tests will hard-fail on non-Triton accelerators (privateuse1/NPU, HPU) whenever CUDA is also present in the process — and the three new classes silently dropped the `is_dynamo_supported()` / `IS_WINDOWS` skips that every other class in the file carries.
> 
> Full review — 3 findings (Infrastructure, Testing, Code Quality)
> ## PR Review: #191577
> ### Summary
> Converts five CUDA-only tests in `test/export/test_serialize.py` into three `instantiate_device_type_tests`-expanded accelerator classes and adds `hw_classification` labels to the existing generic classes. The device-generic direction and the `HardwareClassification` labelling are correct, but the Triton gating regressed in the last two commits and the new classes lost skip decorators, so this needs another pass.
> 
> ### Infrastructure
> **`test/export/test_serialize.py:2645-2646, 2794-2795` — `HAS_TRITON` is a process-level flag, wrong for a device-parameterized class.**
> 
> ```python
> @unittest.skipUnless(HAS_TRITON, "requires triton")
> @skipMPSIf(True, "Triton is not available for MPS")
> def test_triton_hop(self, device) -> None:
> ```
> 
> `HAS_TRITON = has_triton()` (`torch/testing/_internal/inductor_utils.py:60`) answers _"does any device in this process support Triton?"_. `has_triton()` only ever consults `{"cuda", "xpu", "cpu", "mtia"}` (`torch/utils/_triton.py:218-223`) — `privateuse1` and `hpu` are not in that set. In a class expanded per device this produces two wrong outcomes:
> 
> 1. **On a privateuse1-only machine `HAS_TRITON` is `False`**, so `TestSerializeAcceleratorPRIVATEUSE1.test_triton_hop` is skipped. The PR's stated goal — Triton serialize coverage for NPU — yields zero tests.
> 2. **On a machine with both CUDA and a registered privateuse1 backend, `HAS_TRITON` is `True`**, so `TestSerializeAcceleratorPRIVATEUSE1.test_triton_hop` is _not_ skipped and will try to launch a Triton kernel on the privateuse1 device — a hard error, not a skip. The same applies to HPU, which `get_desired_device_type_test_bases` appends unconditionally whenever `TEST_HPU` is set (`torch/testing/_internal/common_device_type.py:1037-1038`), independent of `allow_mps` / `allow_xpu`.
> 
> The `@skipMPSIf(True, ...)` line is evidence of the problem rather than a fix: it hand-patches one instance of "this device has no Triton" while leaving privateuse1 and HPU exposed. As a bonus, PyTorch already has `skipMPS` (`common_device_type.py:2357`) which is literally `skipMPSIf(True, ...)`, so the explicit-`True` form is non-idiomatic too.
> 
> The per-device check you had in commits `09742c3d` / `0258321a` was the right one and should come back:
> 
> ```python
> from torch._dynamo.device_interface import get_interface_for_device
> ...
>     def test_triton_hop(self, device) -> None:
>         if triton is None:
>             self.skipTest("triton package not available")
>         get_interface_for_device(device).raise_if_triton_unavailable(device)
> ```
> 
> `raise_if_triton_unavailable` is the established per-device hook — it's what Inductor's own codegen uses (`torch/_inductor/codegen/wrapper.py:3494`), and every `DeviceInterface` implements it (`torch/_dynamo/device_interface.py:161, 280, 374, 461, 525`). Folding the tests into `TestSerializeAccelerator` (the change @fffrog asked for) is fully compatible with keeping a per-device gate; only the `HAS_TRITON` substitution needs undoing. Once the gate is per-device, both `@unittest.skipUnless(HAS_TRITON, ...)` and `@skipMPSIf(True, ...)` can go, along with the `HAS_TRITON` import at line 59.
> 
> Relatedly, the module-level import guard (lines 63-71) and the skip condition are computed from two independent sources. If `triton` imports but `from torch.library import wrap_triton` raises, the `except ImportError` sets `triton = None` while `HAS_TRITON` stays `True`, and `@triton.jit` inside the test body raises `AttributeError` instead of skipping. Deriving the gate from one source removes that whole class of mismatch.
> 
> [Fix this →](https://claude.ai/code?q=In%20test%2Fexport%2Ftest_serialize.py%20on%20branch%20npu-device-agnostic-test-serialize%2C%20replace%20the%20process-level%20Triton%20gate%20on%20test_triton_hop%20(line%202645)%20and%20test_triton_constexpr_matching%20(line%202794)%20with%20a%20per-device%20gate.%20Remove%20the%20%40unittest.skipUnless(HAS_TRITON%2C%20%22requires%20triton%22)%20and%20%40skipMPSIf(True%2C%20...)%20decorators%20from%20both%20methods%2C%20remove%20the%20now-unused%20HAS_TRITON%20import%20from%20torch.testing._internal.inductor_utils%20at%20line%2059%2C%20and%20instead%20at%20the%20top%20of%20each%20test%20body%20do%3A%20skip%20via%20self.skipTest%20if%20the%20module-level%20triton%20is%20None%2C%20then%20call%20torch._dynamo.device_interface.get_interface_for_device(device).raise_if_triton_unavailable(device).%20Reason%3A%20HAS_TRITON%20is%20has_triton()%20which%20only%20considers%20cuda%2Fxpu%2Fcpu%2Fmtia%2C%20so%20on%20a%20box%20with%20both%20CUDA%20and%20a%20privateuse1%20or%20HPU%20backend%20the%20generated%20PRIVATEUSE1%2FHPU%20test%20classes%20are%20not%20skipped%20and%20hard-fail%20launching%20a%20Triton%20kernel%20on%20a%20device%20that%20has%20no%20Triton.&repo=pytorch/pytorch)
> 
> ### Testing
> **`test/export/test_serialize.py:2615, 2851, 2875` — the three new classes dropped the guard decorators every other class in the file carries.**
> 
> ```python
> class TestSerializeAccelerator(TestCase):        # line 2615 - no decorators
> class TestDeserializeAccelerator(TestCase):      # line 2851 - no decorators
> class TestSaveLoadAccelerator(TestCase):         # line 2875 - no decorators
> ```
> 
> Compare `TestSerialize` (line 105), `TestDeserialize` (943-944), `TestSchemaVersioning` (1892), `TestSaveLoad` (1921), `TestSerializeCustomClass` (2245), `TestPredispatchSerialization` (2524) — all six carry `@unittest.skipIf(not torchdynamo.is_dynamo_supported(), "dynamo doesn't support")`.
> 
> This is not theoretical: `is_dynamo_supported()` returns `False` on Python 3.15+ and on GIL-disabled builds of Python < 3.13.3 (`torch/_dynamo/eval_frame.py:1715-1727`). On those configs these five tests will call `torch.export.export` and error out rather than skip, while their siblings skip cleanly.
> 
> Separately, `test_device` came out of `TestDeserialize`, which also carries `@unittest.skipIf(IS_WINDOWS, "Windows not supported for this test")` (line 943). `TestDeserializeAccelerator` doesn't, so that skip is silently gone. If it's genuinely no longer needed, drop it deliberately and say so in the PR description; don't lose it as a side effect of the move.
> 
> [Fix this →](https://claude.ai/code?q=In%20test%2Fexport%2Ftest_serialize.py%20on%20branch%20npu-device-agnostic-test-serialize%2C%20add%20%40unittest.skipIf(not%20torchdynamo.is_dynamo_supported()%2C%20%22dynamo%20doesn't%20support%22)%20to%20the%20three%20new%20classes%20TestSerializeAccelerator%20(line%202615)%2C%20TestDeserializeAccelerator%20(line%202851)%20and%20TestSaveLoadAccelerator%20(line%202875)%2C%20matching%20every%20other%20test%20class%20in%20the%20file%20(lines%20105%2C%20943-944%2C%201892%2C%201921%2C%202245%2C%202524).%20Also%20add%20%40unittest.skipIf(IS_WINDOWS%2C%20%22Windows%20not%20supported%20for%20this%20test%22)%20to%20TestDeserializeAccelerator%2C%20since%20its%20test_device%20method%20was%20moved%20out%20of%20TestDeserialize%20which%20carries%20that%20skip%20at%20line%20943.%20Reason%3A%20is_dynamo_supported()%20returns%20False%20on%20Python%203.15%2B%20and%20GIL-disabled%20Python%20%3C%203.13.3%2C%20so%20without%20the%20decorator%20these%20tests%20error%20instead%20of%20skipping.&repo=pytorch/pytorch)
> 
> ### Code Quality
> **`test/export/test_serialize.py:2854-2855` — unbound method copying instead of a shared base.**
> 
> ```python
> class TestDeserializeAccelerator(TestCase):
>     _check_graph_nodes = TestDeserialize._check_graph_nodes
>     check_graph = TestDeserialize.check_graph
> ```
> 
> This happens to work today: both helpers only touch `self.assertEqual` / `self.assertTrue` and recurse into `self._check_graph_nodes` (lines 952-1112), all of which resolve on the new class. But the copy takes the two methods and nothing else — in particular not `TestDeserialize.setUp` (lines 948-950), which calls `init_torchbind_implementations()`. `test_device` (a `Conv2d`+`ReLU`+`mul` model) doesn't need torchbind, so it passes, but the next accelerator test that exercises a torchbind op through `check_graph` will fail in a way that gives no hint about the missing `setUp`.
> 
> The state dependency should be explicit rather than inherited-by-copy. Extract the two helpers into a mixin used by both classes:
> 
> ```python
> class _CheckGraphMixin:
>     def _check_graph_nodes(self, gm1, gm2, _check_meta=True): ...
>     def check_graph(self, fn, inputs, ...): ...
> 
> class TestDeserialize(_CheckGraphMixin, TestCase):
>     def setUp(self):
>         super().setUp()
>         init_torchbind_implementations()
> 
> class TestDeserializeAccelerator(_CheckGraphMixin, TestCase):
>     def setUp(self):
>         super().setUp()
>         init_torchbind_implementations()
> ```
> 
> I could not find this class-body-assignment pattern used anywhere else in `test/` for sharing check helpers; mixins and shared bases are the convention.
> 
> [Fix this →](https://claude.ai/code?q=In%20test%2Fexport%2Ftest_serialize.py%20on%20branch%20npu-device-agnostic-test-serialize%2C%20replace%20the%20unbound%20method%20copying%20at%20lines%202854-2855%20(_check_graph_nodes%20%3D%20TestDeserialize._check_graph_nodes%20and%20check_graph%20%3D%20TestDeserialize.check_graph)%20with%20a%20proper%20mixin.%20Extract%20_check_graph_nodes%20(lines%20952-1025)%20and%20check_graph%20(lines%201027-1112)%20out%20of%20TestDeserialize%20into%20a%20module-level%20_CheckGraphMixin%20class%2C%20then%20have%20both%20TestDeserialize%20and%20TestDeserializeAccelerator%20inherit%20from%20it%20alongside%20TestCase.%20Give%20TestDeserializeAccelerator%20the%20same%20setUp%20that%20calls%20init_torchbind_implementations()%20that%20TestDeserialize%20has%20at%20lines%20948-950%2C%20since%20the%20current%20copy%20silently%20omits%20it.&repo=pytorch/pytorch)
> 
> ### Recommendation
> **Request Changes**
> 
> The Infrastructure finding is blocking: as written, the two Triton tests give zero privateuse1 coverage on an NPU-only box and hard-fail on a mixed CUDA+privateuse1 or CUDA+HPU box — the opposite of the PR's goal. The Testing finding is a small, mechanical fix. The mixin change is a cleanup that also removes a real footgun for whoever extends `TestDeserializeAccelerator` next.
> 
> Worth noting for reviewers: commits `09742c3d` and `0258321a` already had the correct per-device gate. The subsequent switch to `HAS_TRITON` was made while addressing the "fold these into `TestSerializeAccelerator`" review comment, but the two changes are independent — the folding can stay, the gate should revert.
> 
> | Branch: [`npu-device-agnostic-test-serialize`](https://github.com/pytorch/pytorch/tree/npu-device-agnostic-test-serialize)

Thanks for the review.

On HAS_TRITON: keeping the current gate. @fffrog asked to fold the Triton cases into TestSerializeAccelerator with @unittest.skipUnless(HAS_TRITON, "requires triton"), and that is the scheme this PR is following.

On the dropped skips: TestDeserializeAccelerator now has the same @skipIf(IS_WINDOWS, ...) as TestDeserialize. All three new accelerator classes also have @skipIf(not torchdynamo.is_dynamo_supported(), ...).

On the helper alias: _check_graph_nodes / check_graph now live on _DeserializeGraphMixin. TestDeserialize and TestDeserializeAccelerator both inherit it, and TestDeserializeAccelerator.setUp calls init_torchbind_implementations() the same way TestDeserialize does.

**Comment by SHIHAIFENG15** [comment]
@can-gaa-hou Claude's review comments have been addressed; the changes are as described in the reply above.
